### ISSUE TITLE
[core] Refine command and state types.

### ISSUE THREAD

**Issue by ejgallego** [Deducteam/lambdapi#87]
As discussed in person, we'd like to have a more refined command type as to be able to properly execute proofs step-by-step in the LSP server having static guarantees.

The rough idea is to refine the command type to:
```ocaml
type cmd_type = WithProof | NoProof

type 'a cmd =
| Theorem : body -> .... -> WithProof cmd
| Definition : body -> .... -> NoProof cmd

type 'a st =
| WithProof :: sign * proof -> Withproof st
| ....

val eval : 'a cmd -> 'a st -> 'a result
val eval_proof : proof_st -> proof_cmd -> proof_st
```

etc... You get the idea; the above needs more refinement of course and a way to properly return the proof commands continuation.

**Comment by rlepigre** [comment]
OK, I'll try to do that as soon as possible!

**Comment by rlepigre** [comment]
@ejgallego: I have tried a couple of things that I don't really find satisfactory. One thing that I would like is to allow the use of some toplevel commands in proofs (`infer` for example). This tends to indicate that we want to mix the type of commands with the type of tactics using a GADT. For example, we would have:
```ocaml
type _ p_item =
  | Symbol : ... -> cmd p_item
  ...
  | Simpl  : ... -> tac p_item
  ...
  | Infer  : ... -> _   p_item

type tactic  = tac p_item Pos.loc
type command = cmd p_item Pos.loc
```

Although doing this simplifies the problem of having two levels of handling, this also means that the handling of an item will possibly yield new items. In particular, a theorem item will add the list of tactics it contains at the head of the list of commands.

What do you think about that?

An alternative would be to make the document aware of the structure (with subcommands). Thus the handling function would give back the list of tactics in the case of `Theorem`.

**Comment by rlepigre** [comment]
OK, so I actually went with the second approach, which seems to make more sense to me. For instance, a theorem is only completely checked when all the tactics have been handle, and they can really be seen as subcommands. In particular, their position is contained in that of the theorem. I pushed the result to the `interactive_proof` branch (see https://github.com/rlepigre/lambdapi/tree/interactive_proof). Let me know what you think!

Note that I did not do the work on the server side (only made sure that things compile), as I do not know what the right approach is on this side. I can probably do it if you tell me how to proceed.

**Comment by rlepigre** [comment]
@ejgallego: any progress on that front?

**Comment by ejgallego** [comment]
Not yet sorry; I'm away for the week, will try to look at this ASAP.

**Comment by rlepigre** [comment]
Fixed by #138.

**Comment by ejgallego** [comment]
I was indeed going to close this too; however note that I still believe that the GADT approach is better in the long term; and more importantly guarantees that `Pure` is correct by construction which is IMHO extremely important.
